### DISCUSSION TITLE
useFieldArray without known path

### DISCUSSION THREAD

**Discussion by bjrice88** [react-hook-form/react-hook-form#7569] [Q&A]
I'm trying to use react-hook-form with a set of somewhat complex forms and an issue I have been running into specifically as it relates to typescript is that I have a defined Array that can appear in a few different paths in the form depending on how it's setup.  

I'm trying to write a component that generalizes using the array but I can't seem to come up with way to both use the `useFieldArray` hook and also pass the fields variable as a prop and correctly define the types when I don't have a specific Array Path.

Is this possible to do without just casting these things as `any` types?

Thanks.

**Comment by Moshyfawn** [comment]
Can you elaborate or phrase differently? An example?
Is #7567 related? Maybe #4055?

**Reply by bjrice88** [reply]
Sure, I know that the question is pretty vague but examples get pretty complicated.

So in my situation let's say my defaultValues look like this.
```
const defaultValues = {
  test: [
    {
      name: "useFieldArray1",
      nestedArray: [{ field1: "field1", field2: "field2" }]
    },
    {
      name: "useFieldArray2",
      nestedArray: [{ field1: "field1", field2: "field2" }]
    }
  ],
  nestedArray: [{ field1: "field1", field2: "field2" }]
};
```

I want to write one component that will handle both `nestedArray` inputs.  I'm pretty sure the way to handle that would be to pass in the path to the array as a prop to the component.  Something like this.

```
export default ({ arrayPath, control, register }) => {
  const { fields, remove, append } = useFieldArray({
    control,
    name: `${nestIndex}`
  });
```

The problem I'm having is that according to the docs you have to cast the path so it looks like this.
```
export default ({ nestIndex, control, register }) => {
  const { fields, remove, append } = useFieldArray({
    control,
    name: `${nestIndex}`  as `test.0.nestedArray`
  });
```

But then I would need another component with the name cast as just `nestedArray`  And because this is just a simplified version of what I mean, there are a dynamic number of places that this component could be used in the overall form.

I know this is partly just because TypeScript is definitely not my strong suit but I can manage normally.

The related problem is just that let's say in the above example I want to pass `fields` to another component as a prop.  The only way I have found to do that is to type the prop like this.  `fields: FieldArrayWithId<FormInputs, "test.0.nestedArray", "id">[];` but I have the same problem here where the path might not be statically known and I'm not sure how to handle that situation in TypeScript.

Thanks!

**Comment by Moshyfawn** [comment]
#7354 might actually be what you're looking for. [Next RHF version](https://github.com/react-hook-form/react-hook-form/discussions/7433) is gonna be focused on improving generic types besides anything else

**Reply by bjrice88** [reply]
It is indeed, I actually found that earlier today searching around.  I'm just glad that I'm not off my rocker trying to figure it out.  It does appear that there is a workaround with type assertion that I'll need to look into though.
